### PR TITLE
add mutiprocess role maker so that users can easily switch single gpu…

### PR DESCRIPTION
**Pull request by guru4elephant**
… card to multi gpu card with fleet api

test=develop

With new apis, a user can switch from single gpu card to multi gpu card on a single machine with fleet API. An single gpu card example is as follows:

```
import paddle.fluid as fluid

# you should define a loss by a neural nets, and here is just an example
total_loss = fluid.layers.data(name="loss", shape=[1])
role = role_maker.MultiProcessRoleMaker()
optimizer = fluid.optimizer.Adam(learning_rate=0.01)
opts, param_and_grads = optimizer.minimize(total_loss)

```

A multi process gpu training program is as follows:

```
import paddle.fluid as fluid
from paddle.fluid.incubate.fleet.collective import fleet, DistributedStrategy
import paddle.fluid.incubate.fleet.base.role_maker as role_maker

# you should define a loss by a neural nets, and here is just an example
total_loss = fluid.layers.data(name="loss", shape=[1])
role = role_maker.MultiProcessRoleMaker()
fleet.init(role)
optimizer = fluid.optimizer.Adam(learning_rate=0.01)
optimizer = fleet.distributed_optimizer(optimizer)
opts, param_and_grads = optimizer.minimize(total_loss)

```

A start command of multi-process gpu training is 
```
python -m paddle.distributed.launch --gpus 8 my_distriburted_trainer.py
```

### CODE REVIEW

File: python/paddle/fluid/incubate/fleet/base/fleet_base.py [hunk 1]
```diff
@@ -189,18 +189,8 @@ def init(self, role_maker=None):
...
-        elif isinstance(role_maker, UserDefinedRoleMaker):
-            self._role_maker = role_maker
-
-        else:
-            raise ValueError(
-                "role_maker must be an instance of UserDefinedRoleMaker/MPISymetricRoleMaker"
-            )
-
+        self._role_maker = role_maker
+        self._role_maker.generate_role()
```
Comment (sandyhouse) [inline python/paddle/fluid/incubate/fleet/base/fleet_base.py:193]: Not all RoleMaker have the method 'generate_role()'

File: python/paddle/fluid/incubate/fleet/collective/__init__.py [hunk 2]
```diff
@@ -69,6 +125,85 @@ def save_persistables(self, executor, dirname, main_program=None):
...
+
+    def _transpile_program(self, startup_program=None):
+        startup_program = startup_program if startup_program else \
+                          fluid.framework.default_startup_program()
+        worker_endpoints = fleet.worker_endpoints()
+        trainer_id = fleet.worker_index()
+        current_endpoint = fleet.worker_endpoints()[trainer_id]
+        # call transpiler
+        config = dist_transpiler.DistributeTranspilerConfig()
+        config.mode = "collective"
```
Comment (sandyhouse) [inline python/paddle/fluid/incubate/fleet/collective/__init__.py:144]: should be config.mode='nccl2'?

File: python/paddle/fluid/incubate/fleet/base/fleet_base.py (hunk 1 above)
Comment (guru4elephant) [reply python/paddle/fluid/incubate/fleet/base/fleet_base.py:193]: generate_role() is base class method that can be called. default implementation is do nothing.

File: python/paddle/fluid/incubate/fleet/collective/__init__.py (hunk 2 above)
Comment (gavin1332) [reply python/paddle/fluid/incubate/fleet/collective/__init__.py:144]: collective config is not in develop branch

Comment (guru4elephant) [reply python/paddle/fluid/incubate/fleet/collective/__init__.py:144]: yes, I should wait the collective mode transpiler to merge

**Pull request closed without merging**